File: Python/Programmars/Level_2/메뉴 리뉴얼.py
from collections import Counter
from itertools import combinations


def solution(orders, course):
    menus = dict()
    answer = []
    for x in course:
        for order in orders:
            for y in combinations(sorted(order), x):
                menus[y] = menus.get(y, 0) + 1
    for x in course:
        max_menu = 0
        for menu, count in menus.items():
            if len(menu) == x:
                max_menu = max(max_menu, count)
        for menu, count in menus.items():
            if max_menu >= 2 and len(menu) == x and count == max_menu:
                answer.append(''.join(menu))
    return sorted(answer)


# ["AC", "ACDE", "BCFG", "CDE"]
print(solution(["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"], [2, 3, 4]))

# ["ACD", "AD", "ADE", "CD", "XYZ"]
print(solution(["ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"], [2, 3, 5]))

print(solution(["XYZ", "XWY", "WXA"], [2, 3, 4]))  # ["WX", "XY"]


# ref
def solution(orders, course):
    result = []

    for course_size in course:
        order_combinations = []
        for order in orders:
            order_combinations += combinations(
                sorted(order), course_size)

        most_ordered = Counter(order_combinations).most_common()
        result += [k for k, v in most_ordered if v >
                   1 and v == most_ordered[0][1]]

    return [''.join(v) for v in sorted(result)]

File: JavaScript/Programmars/카카오블라인드/메뉴 리뉴얼.js
// https://programmers.co.kr/learn/courses/30/lessons/72411?language=javascript
const solution = (orders, course) => {
  const answer = [];
  const orderArr = orders.map((order) => order.split('').sort());
  let courseObj, courseLen, temp, menus, menuLen, maxCourseCount, result;

  const combination = (level, start) => {
    if (level === courseLen) {
      result = temp.join('');
      if (courseObj[result]) {
        courseObj[result]++;
        maxCourseCount = Math.max(maxCourseCount, courseObj[result]);
      } else courseObj[result] = 1;
    } else {
      for (let i = start; i < menuLen; i++) {
        temp[level] = menus[i];
        combination(level + 1, i + 1);
      }
    }
  };

  for (let num of course) {
    courseObj = {};
    courseLen = num;
    temp = Array.from({ length: num }, () => 0);
    maxCourseCount = 1;
    for (let order of orderArr) {
      menus = order;
      menuLen = menus.length;
      combination(0, 0);
    }

    // maxCourseCount를 따로 배열에 저장해서 최상위 스코프에서 filter로 answer 만드는 게 더 효율적
    if (maxCourseCount < 2) continue;
    for (let [courseStr, courseCount] of Object.entries(courseObj)) {
      if (courseCount === maxCourseCount) {
        answer.push(courseStr);
      }
    }
  }

  answer.sort();

  return answer;
};

console.log(
  solution(['ABCFG', 'AC', 'CDE', 'ACDE', 'BCFG', 'ACDEH'], [2, 3, 4])
); // 	["AC", "ACDE", "BCFG", "CDE"]
console.log(
  solution(['ABCDE', 'AB', 'CD', 'ADE', 'XYZ', 'XYZ', 'ACD'], [2, 3, 5])
); // ["ACD", "AD", "ADE", "CD", "XYZ"]
console.log(solution(['XYZ', 'XWY', 'WXA'], [2, 3, 4])); // ["WX", "XY"]
